urllib3/util/hpkp.py

```diff
@@ -0,0 +1,336 @@
+        address.
+    """
+    # TODO: This method is super complicated, refactor.
+    # Parse the directives. Split any OWS, and then split the directives.
+    directives = header.split(';')
+
+    pin_directives = []
+    max_age = None
+    include_subdomains = False
+    report_uri = None
+
+    for directive in directives:
+        # Strip any OWS, then split into name and value. All directives have
+        # names, but frustratingly they don't all have values.
+        directive = directive.strip().split('=', 1)
+        name = directive[0].lower()
+        try:
+            value = directive[1]
+        except IndexError:
+            value = None
+
+        if name == 'pin-sha256':
+            value = unquote_string(value)
+            pin_directives.append(value)
+        elif name == 'max-age':
+            # There may only be one max-age directive. Be careful about
+            # policing this, we don't want to accidentally allow more than one.
+            if max_age:
+                raise HPKPError("Multiple max-age directives in PKP header")
+
+            max_age = int(unquote_string(value))
+            if max_age < 0:
+                raise HPKPError("max-age must be positive")
+        elif name == 'includesubdomains':
+            include_subdomains = True
+        elif name == 'report-uri':
+            if report_uri:
+                raise HPKPError("Multiple report-uri directives in PKP header")
+
+            report_uri = unquote_string(value)
+
+    if not pin_directives:
+        raise HPKPError("No pin directives in PKP header")
+
+    # LUKASA: Is the use of time.time() safe here? Probably not.
+    return KnownPinnedHost(
+        domain, pin_directives, max_age, include_subdomains, report_uri, time.time()
+    )
+
+
+def _validate_pin(connection, host, shortcut=True):
+    """
+    Validates that a TLS connection matches a specific pin.
+
+    If shortcut is ``True``, returns as soon as a matching cert is found: do
+    not use the return value in this case.
+
+    Returns True if the pin contains at least one key *not* used in this
+    certificate chain. Returns False if the pin only contains certificates in
+    the chain.
+    """
+    match = False
+    non_match = False
+
+    # We ideally want to grab the whole cert chain here.
+    certificates = connection.sock.get_peer_cert_chain()
+
+    for binary_certificate in certificates:
+        # For each cert in the chain, get a base64-encoded form of the
+        # SHA256 of the public key and check whether it's in the pin list.
+        if _certificate_in_pins(binary_certificate, host):
+            if shortcut:
+                return True
+
+            match = True
+        else:
+            non_match = True
+
+    if not match:
+        raise HPKPError("Failed to validate trust chain with HPKP!")
+
+    return match and non_match
+
+
+def _certificate_in_pins(der_certificate, host):
```

Comment:
Maybe this should be split into `pubkey_fingerprint(der_cert, hash=hashlib.sha256)`
and a function which actually does the verification.